pkg/conversion/converter.go

```diff
@@ -0,0 +1,245 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package conversion
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/google/cel-go/interpreter"
+
+	apiextensionsinternal "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	structuralschema "k8s.io/apiextensions-apiserver/pkg/apiserver/schema"
+	structuraldefaulting "k8s.io/apiextensions-apiserver/pkg/apiserver/schema/defaulting"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+)
+
+func NewConverter(
+	crd *apiextensionsv1.CustomResourceDefinition,
+	apiConversion *apisv1alpha1.APIConversion,
+) (*Converter, error) {
+	structuralSchemas := make(map[string]*structuralschema.Structural)
+
+	// Gather all the structural schemas for easy map-based indexing
+	for _, v := range crd.Spec.Versions {
+		if v.Schema == nil {
+			continue
+		}
+
+		internalJSONSchemaProps := &apiextensionsinternal.JSONSchemaProps{}
+		if err := apiextensionsv1.Convert_v1_JSONSchemaProps_To_apiextensions_JSONSchemaProps(v.Schema.OpenAPIV3Schema, internalJSONSchemaProps, nil); err != nil {
+			return nil, fmt.Errorf("failed converting version %s validation to internal version: %w", v.Name, err)
+		}
+
+		structuralSchema, err := structuralschema.NewStructural(internalJSONSchemaProps)
+		if err != nil {
+			return nil, fmt.Errorf("error getting structural schema for version %s: %w", v.Name, err)
+		}
+
+		if err := structuraldefaulting.PruneDefaults(structuralSchema); err != nil {
+			return nil, fmt.Errorf("error pruning defaults for version %s: %w", v.Name, err)
+		}
+
+		structuralSchemas[v.Name] = structuralSchema
+	}
+
+	compiledRules, err := CompileConversions(apiConversion, structuralSchemas)
+	if err != nil {
+		return nil, fmt.Errorf("error compiling conversion rules: %w", err)
+	}
+
+	return &Converter{
+		crd:               crd,
+		compiledRules:     compiledRules,
+		structuralSchemas: structuralSchemas,
+		apiConversion:     apiConversion,
+	}, nil
+}
+
+type Converter struct {
+	crd               *apiextensionsv1.CustomResourceDefinition
+	compiledRules     map[string][]*CompiledRule
+	structuralSchemas map[string]*structuralschema.Structural
+	apiConversion     *apisv1alpha1.APIConversion
+}
+
+func (c *Converter) Convert(list *unstructured.UnstructuredList, targetGV schema.GroupVersion) (*unstructured.UnstructuredList, error) {
```

Comment:
not sure that is documented elsewhere: the input list is never mutated?

Reply:
It would not be good if the input list were mutated. There's no guard that prevents it currently. I might need to make a slight change to the upstream code:

```go
convertedObjects, err := c.converter.Convert(list, toGVK.GroupVersion())
// error checking

convertedList := list.DeepCopy()
// loop through convertedObjects and assign each one to the appropriate index in 'list'
```

If `list` is modified by `c.converter.Convert`, results are not guaranteed. I should move the DeepCopy up.

Reply:
https://github.com/kubernetes/kubernetes/pull/114959, will pull into our fork